toml/toml.go

```diff
@@ -89,3 +91,51 @@ func (s *Size) UnmarshalText(text []byte) error {
 	*s = Size(size)
 	return nil
 }
+
+type FileMode uint32
+
+func (m *FileMode) UnmarshalText(text []byte) error {
+	// Ignore if there is no value set.
+	if len(text) == 0 {
+		return nil
+	}
+
+	if text[0] != '0' {
+		return errors.New("the first character of an octal must be zero")
+	}
+	mode, err := strconv.ParseUint(string(text[1:]), 8, 32)
+	if err != nil {
+		return err
+	}
+	*m = FileMode(mode)
+	return nil
+}
+
+func (m FileMode) MarshalText() (text []byte, err error) {
+	if m != 0 {
+		return []byte(fmt.Sprintf("0%03o", m)), nil
```

Comment:
Could be `%04o`?

Reply:
Is there a reason for `%04o`?

Reply:
One thing I just noticed is that the output from `MarshalText` appears to be different than what `UnmarshalText` parses so I'll update that.

Reply:
`%03o` sets the output width to three and left pads with `0`s. The leading `0` could be be part of the padding, i.e. `%04o` should be equivalent to `0%03o` which would be the same as `00%02o` etc., as long as the value `m` didn't exceed the given width.

Reply:
I think it would be safer to explicitly put the zero. It's possible for a file to have the mode `1777` because of things like the [sticky bit](https://en.wikipedia.org/wiki/Sticky_bit).

On the other hand, the value here is always going to be an octal so we could go and declare that `777` will always be interpreted as an octal, even if there is no leading zero. Then we would use `%04` instead (and get rid of the trailing `o` since it's just confusing).

Reply:
The trailing zero is to tell `Sprintf` to use octal and I just remembered that so just ignore that part. We would use `%04o`.